### PR TITLE
Implement reporting DSO names from PyPerf

### PR DESCRIPTION
**Pull request by marcin-ol**
Support reporting DSO name for native frames.
Can be triggered with flag `--do-insert-dso-name`.

### CODE REVIEW

**Comment by marcin-ol** [comment]
DSO reporting is working fine in the following cases:
- regular python process, with shared object dependencies loaded at intialization,
- process replacing the executable image with `exec` syscall.

After first conversion of native stack, a cache of symbols is maintained by `ProcSyms` object.
Cache needs to be refreshed in case a process loads shared objects dynamically (which happens in case of modules with native code). 
Refresh mechanism will be implemented in upcoming commits.

File: examples/cpp/pyperf/PyPerfProfiler.cc [hunk 1]
```diff
@@ -256,10 +262,8 @@ bool PyPerfProfiler::populatePidTable() {
       continue;
     }
     pid_config_map.update_value(pid, pidData);
-
```
Comment (Jongy) [inline examples/cpp/pyperf/PyPerfProfiler.cc:259]: Can you please undo unrelated changes? (blank additions/removals in unrelated code)

File: examples/cpp/pyperf/PyPerfProfiler.h [hunk 2]
```diff
@@ -67,7 +67,8 @@ class PyPerfProfiler {
   // init must be invoked exactly once before invoking profile
   PyPerfResult init(unsigned int symbolsMapSize, unsigned int eventsBufferPages,
                     unsigned int kernelStacksMapSize, unsigned int userStacksPages,
-                    unsigned int fsOffset, unsigned int stackOffset);
+                    unsigned int fsOffset, unsigned int stackOffset,
+                    bool insertDsoName=true);
```
Comment (Jongy) [inline examples/cpp/pyperf/PyPerfProfiler.h:71]: Any reason we add a default value for it? The c'tor is invoked only at one site afaik

File: examples/cpp/pyperf/PyPerfProfiler.cc [hunk 3]
```diff
@@ -230,24 +234,26 @@ Populate the PidData BPF map.
...
   for (const auto pid : pid_config_keys) {
     auto pos = std::find(pids.begin(), pids.end(), pid);
     if (pos == pids.end()) {
+      // Remove dead pid from config map and native stack trace's cache
       pid_config_map.remove_value(pid);
+      NativeStackTrace::prune_dead_pid(pid);
     }
     else {
       result = true;
+      // To avoid re-population
```
Comment (Jongy) [inline examples/cpp/pyperf/PyPerfProfiler.cc:250]: Don't we ever `update_value` of existing PIDs?

I looked at the fields:
```
typedef struct pid_data {
  enum pthreads_impl pthreads_impl;
  struct py_globals globals;
  struct struct_offsets offsets;
  uintptr_t interp;  // vaddr of PyInterpreterState
} PidData;
```

from a skim on `tryTargetPid`, it seems to work all-or-nothing, so if it returned `true` we know all fields are filled in, and we won't need re-update because those fields don't change for the same process. Did you consider it further? If not, we need to inspect if this logic change makes sense.

Btw what caused you to make the change, anyway? (it appears to be a bug in the logic but I wonder how you arrived at it :thinking: )

**Review by Jongy — COMMENTED** [review]
Not done reviewing

File: examples/cpp/pyperf/PyPerfProfiler.cc (hunk 1 above)
Comment (marcin-ol) [reply examples/cpp/pyperf/PyPerfProfiler.cc:259]: 👍

File: examples/cpp/pyperf/PyPerfProfiler.h (hunk 2 above)
Comment (marcin-ol) [reply examples/cpp/pyperf/PyPerfProfiler.h]: No, no default value needed. Thanks for spotting this.

File: examples/cpp/pyperf/PyPerfNativeStackTrace.cc [hunk 4]
```diff
@@ -200,5 +226,36 @@ bool NativeStackTrace::error_occured() const {
...
+    procSymbolsCache.erase(it);
+  }
+}
+
+void NativeStackTrace::enable_dso_reporting() {
+  NativeStackTrace::insert_dso_name = true;
+}
+
+ProcSyms* NativeStackTrace::get_proc_symbols(uint32_t pid) {
+  static struct bcc_symbol_option symbol_options = {
```
Comment (Jongy) [inline examples/cpp/pyperf/PyPerfNativeStackTrace.cc:241]:
```suggestion
  static const struct bcc_symbol_option symbol_options = {
```

Comment (Jongy) [reply examples/cpp/pyperf/PyPerfNativeStackTrace.cc:241]: Or actually, it seems that this function doesn't get `const`. Oh well

File: examples/cpp/pyperf/PyPerfNativeStackTrace.cc [hunk 5]
```diff
@@ -200,5 +226,36 @@ bool NativeStackTrace::error_occured() const {
...
+    .use_debug_file = 1,
+    .check_debug_file_crc = 1,
+    .lazy_symbolize = 1,
+    .use_symbol_type = BCC_SYM_ALL_TYPES
+  };
+  auto current_time_point = std::chrono::steady_clock::now();
+  double timestamp_s = std::chrono::duration_cast<std::chrono::duration<double>>(current_time_point - startup_time_point).count();
```
Comment (Jongy) [inline examples/cpp/pyperf/PyPerfNativeStackTrace.cc:248]: You don't need to subtract `startup_time_point` - I don't think that the latter is needed at all, just check `now() - procSymbolsCache[pid].timestamp_s > ProcSymbolsCacheTTL_S` and that should be enough, no?

File: examples/cpp/pyperf/PyPerfProfiler.cc (hunk 1 above)
Comment (Jongy) [reply examples/cpp/pyperf/PyPerfProfiler.cc:259]: (was not pushed yet, leaving this unresolved)

File: examples/cpp/pyperf/PyPerfNativeStackTrace.cc [hunk 6]
```diff
@@ -65,23 +81,33 @@ NativeStackTrace::NativeStackTrace(uint32_t pid, const unsigned char *raw_stack,
...
     // TODO: This function is very heavy. We should try to do some caching here, maybe in the
     //       underlying UPT function.
     res = unw_get_proc_name(&cursor, sym, sizeof(sym), &offset);
     if (res == 0) {
-      realname = abi::__cxa_demangle(sym, NULL, NULL, &status);
-      if (status == 0) {
-        this->symbols.push_back(std::string(realname));
+      unw_word_t ip;
+      unw_get_reg(&cursor, UNW_REG_IP, &ip);
+      if (procSymbols && procSymbols->resolve_addr(ip, &resolved_symbol, true)) {
```
Comment (Jongy) [inline examples/cpp/pyperf/PyPerfNativeStackTrace.cc:97]: Let me make sure I understand the logic (and in any case, please document it here). What we do now is:
1. Unwind via libunwind
2. Per IP, if DSO names are enabled, we attempt to symbolicate with `ProcSyms`. If it succeeds we add a symbol with the demangled name + DSO name.
3. Per IP, if DSO names are disabled OR `ProcSyms` symbolication failed, we fall back to the old method of libunwind + `__cxa_demangle`.
Correct?

File: examples/cpp/pyperf/PyPerfNativeStackTrace.cc [hunk 7]
```diff
@@ -65,23 +81,33 @@ NativeStackTrace::NativeStackTrace(uint32_t pid, const unsigned char *raw_stack,
...
+        char* demangled = nullptr;
+        demangled = abi::__cxa_demangle(sym, nullptr, nullptr, &status);
+        if (!status) {
+          this->symbols.push_back(std::string(demangled));
+          free(demangled);
+        } else {
+          this->symbols.push_back(std::string(sym));
+        }
       }
-      free(realname);
```
Comment (Jongy) [inline examples/cpp/pyperf/PyPerfNativeStackTrace.cc:84]: Previously we `free`d in all cases, even if `__cxa_demangle` reportedly fails. Now you `free` only in the success path? Any reason for this change?

I don't recall why we originally did it this way - please see in the PR why, explain the reason to change or revert the change such that `free` is always called in the `__cxa_demangle` path.

File: examples/cpp/pyperf/PyPerfNativeStackTrace.cc [hunk 8]
```diff
@@ -65,23 +81,33 @@ NativeStackTrace::NativeStackTrace(uint32_t pid, const unsigned char *raw_stack,
...
+
   do {
     unw_word_t offset;
-    char sym[256];
-    char   *realname;
-    int     status;
-
     // TODO: This function is very heavy. We should try to do some caching here, maybe in the
     //       underlying UPT function.
     res = unw_get_proc_name(&cursor, sym, sizeof(sym), &offset);
```
Comment (Jongy) [inline examples/cpp/pyperf/PyPerfNativeStackTrace.cc:93]: A slight optimization - this can move to the non-DSO branch, because if we use `ProcSyms,` then this result isn't used at all, right? Technically we can unwind with libunwind and symbolicate with BCC.

File: examples/cpp/pyperf/PyPerfNativeStackTrace.cc [hunk 9]
```diff
@@ -200,5 +226,36 @@ bool NativeStackTrace::error_occured() const {
...
+    .lazy_symbolize = 1,
+    .use_symbol_type = BCC_SYM_ALL_TYPES
+  };
+  auto current_time_point = std::chrono::steady_clock::now();
+  double timestamp_s = std::chrono::duration_cast<std::chrono::duration<double>>(current_time_point - startup_time_point).count();
+  if (!procSymbolsCache.count(pid)) {
+    procSymbolsCache[pid] = ProcSymbolsCacheEntry{timestamp_s, unique_ptr<ProcSyms>(new ProcSyms(pid, &symbol_options))};
+  } else {
+    if (timestamp_s - procSymbolsCache[pid].timestamp_s > ProcSymbolsCacheTTL_S) {
+      procSymbolsCache[pid].proc_syms->refresh();
```
Comment (Jongy) [inline examples/cpp/pyperf/PyPerfNativeStackTrace.cc:253]: I looked at its implementation, very inefficient :/ It flushes all loaded modules and re-loads them, instead of removing modules that have changed loading new ones.

Well, we'll talk about performance after CR

**Review by Jongy — CHANGES_REQUESTED** [review]
Done first review

File: examples/cpp/pyperf/PyPerfProfiler.cc (hunk 3 above)
Comment (Jongy) [reply examples/cpp/pyperf/PyPerfProfiler.cc:250]: Hmm, I'm not sure on which part I originally wrote it. I think I might have mis-read the diff and thought that you added the `pids.erase(pos)` call in this PR? I now see that you didn't, and the logic appears sound to me (1. get running python pids; 2. prune dead pids from kernel map; 2. prune pids already in kernel map from pids list; 4. adds pythons in pids list to kernel map) . You're welcome to add a bit more of documentation around here as you've added, on the flow of this function, since it's a complex one. This comment is resolved IMO, leaving it open for you to read.

File: examples/cpp/pyperf/PyPerfNativeStackTrace.cc (hunk 6 above)
Comment (marcin-ol) [reply examples/cpp/pyperf/PyPerfNativeStackTrace.cc:97]: Yes, that's right.

File: examples/cpp/pyperf/PyPerf.cc [hunk 10]
```diff
@@ -126,6 +138,7 @@ int main(int argc, char** argv) {
     found = found || parseStrArg({"-o", "--output"}, output);
     found = found || parseIntArg({"--fs-offset"}, fsOffset);
     found = found || parseIntArg({"--stack-offset"}, stackOffset);
+    found = found || parseFlag({"--do-insert-dso-name"}, insertDsoName);
```
[the remark this answers is not in the document]
Comment (marcin-ol) [reply examples/cpp/pyperf/PyPerf.cc:141]: Agreed.

File: examples/cpp/pyperf/PyPerfNativeStackTrace.cc [hunk 11]
```diff
@@ -39,6 +50,11 @@ NativeStackTrace::NativeStackTrace(uint32_t pid, const unsigned char *raw_stack,
   unw_accessors_t my_accessors = _UPT_accessors;
   my_accessors.access_mem = NativeStackTrace::access_mem;
   my_accessors.access_reg = NativeStackTrace::access_reg;
+  char sym[256];
+  ProcSyms* procSymbols = nullptr;
+  // reserve memory for platform-defined path limit AND the symbol
+  size_t buf_size = sizeof(sym) + PATH_MAX + 4;
```
[the remark this answers is not in the document]
Comment (marcin-ol) [reply examples/cpp/pyperf/PyPerfNativeStackTrace.cc:56]: 4 additional characters are due to formatting of DSO name (parens, whitespace and padding to even value).

File: examples/cpp/pyperf/PyPerfNativeStackTrace.cc (hunk 8 above)
Comment (marcin-ol) [reply examples/cpp/pyperf/PyPerfNativeStackTrace.cc:93]: Yes, I believe that should work this way.

File: examples/cpp/pyperf/PyPerfNativeStackTrace.cc (hunk 7 above)
Comment (marcin-ol) [reply examples/cpp/pyperf/PyPerfNativeStackTrace.cc:84]: This change wasn't necessary. I guess I lacked the sense of safety seeing `free()` called when argument might be `NULL`.
I will revert it to the previous state, in the vein of keeping least diff necessary.

Actually that wouldn't be a mistake according to documentation. `__cxa_demangle` would only return actual buffer in the success path. Any other case, and it returns `nullptr`.

File: examples/cpp/pyperf/PyPerfNativeStackTrace.cc (hunk 5 above)
Comment (marcin-ol) [reply examples/cpp/pyperf/PyPerfNativeStackTrace.cc:248]: Good idea. 
Direct cast is not available, but a simple workaround can be built with `time_point::time_since_epoch()`.

File: examples/cpp/pyperf/PyPerfNativeStackTrace.cc (hunk 4 above)
Comment (marcin-ol) [reply examples/cpp/pyperf/PyPerfNativeStackTrace.cc:241]: That would require changing c-tor signature for `ProcSyms`. This could be a part of a larger effort, once we decide we decide improvements are necessary in baseline code.

File: examples/cpp/pyperf/PyPerfNativeStackTrace.cc (hunk 6 above)
Comment (marcin-ol) [reply examples/cpp/pyperf/PyPerfNativeStackTrace.cc]: After latest commit including suggestion https://github.com/Granulate/bcc/pull/38#discussion_r1020957591 the logic is as follows:
1. If DSO names are enabled, attempt to symbolicate with `ProcSyms`. If it succeeds we add a symbol with the demangled name + DSO name.
2. If step 2. failed, or DSO was not enabled, Unwind via `libunwind` + `__cxa_demangle`.

File: examples/cpp/pyperf/PyPerfNativeStackTrace.cc (hunk 8 above)
Comment (marcin-ol) [reply examples/cpp/pyperf/PyPerfNativeStackTrace.cc]: Adopted the suggested optimization.

File: examples/cpp/pyperf/PyPerfProfiler.cc (hunk 1 above)
Comment (Jongy) [reply examples/cpp/pyperf/PyPerfProfiler.cc:259]: @marcin-ol was not pushed yet' let's remove all unrelated formatting fixes in this PR, blank changes etc.

File: examples/cpp/pyperf/PyPerfNativeStackTrace.cc (hunk 11 above)
[the remark this answers is not in the document]
Comment (Jongy) [reply examples/cpp/pyperf/PyPerfNativeStackTrace.cc:59]: Then can be replaced with `sizeof(" ()")` :) makes it clear for the reader.

Why "padding to even value"?

File: examples/cpp/pyperf/PyPerfNativeStackTrace.cc (hunk 4 above)
Comment (Jongy) [reply examples/cpp/pyperf/PyPerfNativeStackTrace.cc:237]: No no, not required. If the function has a bad signature then that's another problem.

File: examples/cpp/pyperf/PyPerfNativeStackTrace.cc (hunk 7 above)
Comment (Jongy) [reply examples/cpp/pyperf/PyPerfNativeStackTrace.cc:84]: If that's the documentation, it's fine. `free(NULL)` is always legit. I wanted to see if there was a reason for it.

Comment (Jongy) [reply examples/cpp/pyperf/PyPerfNativeStackTrace.cc:84]: Resolving this

File: examples/cpp/pyperf/PyPerfNativeStackTrace.cc (hunk 8 above)
Comment (Jongy) [reply examples/cpp/pyperf/PyPerfNativeStackTrace.cc]: Cool

File: examples/cpp/pyperf/PyPerfNativeStackTrace.cc [hunk 12]
```diff
@@ -39,6 +53,11 @@ NativeStackTrace::NativeStackTrace(uint32_t pid, const unsigned char *raw_stack,
...
+  unique_ptr<char[]> buf(new char[buf_size]);
```
Comment (Jongy) [inline examples/cpp/pyperf/PyPerfNativeStackTrace.cc:60]: Btw, any reason not to place it on the stack?

File: examples/cpp/pyperf/PyPerfNativeStackTrace.cc (hunk 9 above)
Comment (Jongy) [reply examples/cpp/pyperf/PyPerfNativeStackTrace.cc:248]: Btw - will it handle if the PID changes? i.e let's say we had python with PID=55. Then 10 minutes passed and now PID=55 is a different Python application. Will `refresh` work fine?

Comment (Jongy) [reply examples/cpp/pyperf/PyPerfNativeStackTrace.cc:248]: (or, do we expect `prune_dead_pid` to prune the dead PIDs before they are being recycled)

File: examples/cpp/pyperf/PyPerfNativeStackTrace.cc (hunk 11 above)
[the remark this answers is not in the document]
Comment (marcin-ol) [reply examples/cpp/pyperf/PyPerfNativeStackTrace.cc]: Good idea with the constant. Not sure why I invented this padding here.
👍

File: examples/cpp/pyperf/PyPerfNativeStackTrace.cc (hunk 12 above)
Comment (marcin-ol) [reply examples/cpp/pyperf/PyPerfNativeStackTrace.cc]: No, I guess it can safely be moved to stack. I will fix this.

File: examples/cpp/pyperf/PyPerfNativeStackTrace.cc (hunk 9 above)
Comment (marcin-ol) [reply examples/cpp/pyperf/PyPerfNativeStackTrace.cc:248]: That's a valid concern.
We might run into 1 minute of stale data, with the current refresh implementation.
`prune_dead_pid` might not help with pid reuse issue.
We might need to research this a bit.
I can think of two approaches:
- improve identification of processes - use more information, perhaps process start_time?, and store it in `PidData`,
- register a process exit hook with eBPF to handle process departure.
The first one might be simpler to take.

File: examples/cpp/pyperf/PyPerfProfiler.cc (hunk 1 above)
Comment (marcin-ol) [reply examples/cpp/pyperf/PyPerfProfiler.cc]: Should be cleaned up now.

File: examples/cpp/pyperf/PyPerfNativeStackTrace.cc (hunk 9 above)
Comment (Jongy) [reply examples/cpp/pyperf/PyPerfNativeStackTrace.cc:248]: The first approach is simpler to take, but anyway - `updateProcesses`'s default interval is 10 seconds, so `populatePidTable` is called every 10s and thus `prune_dead_pid`. I think 10s is good enough - I don't expect PID reuse in such a short timeframe, it's extreme

Comment (Jongy) [reply examples/cpp/pyperf/PyPerfNativeStackTrace.cc:248]: Resolving this

File: examples/cpp/pyperf/PyPerfNativeStackTrace.cc (hunk 6 above)
Comment (Jongy) [reply examples/cpp/pyperf/PyPerfNativeStackTrace.cc]: Kk good

File: examples/cpp/pyperf/PyPerfNativeStackTrace.cc [hunk 13]
```diff
@@ -39,6 +53,11 @@ NativeStackTrace::NativeStackTrace(uint32_t pid, const unsigned char *raw_stack,
   unw_accessors_t my_accessors = _UPT_accessors;
   my_accessors.access_mem = NativeStackTrace::access_mem;
   my_accessors.access_reg = NativeStackTrace::access_reg;
+  char sym[256];
+  ProcSyms* procSymbols = nullptr;
+  // reserve memory for platform-defined path limit AND the symbol
+  size_t buf_size = sizeof(sym) + PATH_MAX + sizeof("() ");
```
Comment (Jongy) [inline examples/cpp/pyperf/PyPerfNativeStackTrace.cc:59]:
```suggestion
  const size_t buf_size = sizeof(sym) + PATH_MAX + sizeof("() ");
```

File: examples/cpp/pyperf/PyPerfNativeStackTrace.cc [hunk 14]
```diff
@@ -65,26 +84,29 @@ NativeStackTrace::NativeStackTrace(uint32_t pid, const unsigned char *raw_stack,
     goto out;
   }
 
+  struct bcc_symbol resolved_symbol;
```
Comment (Jongy) [inline examples/cpp/pyperf/PyPerfNativeStackTrace.cc:87]: Declared here but not used here - declare it inside the loop, where it's actually used?

File: examples/cpp/pyperf/PyPerfNativeStackTrace.cc [hunk 15]
```diff
@@ -65,26 +84,29 @@ NativeStackTrace::NativeStackTrace(uint32_t pid, const unsigned char *raw_stack,
...
+    procSymbols = get_proc_symbols(pid);
+  }
+
   do {
     unw_word_t offset;
-    char sym[256];
-    char   *realname;
-    int     status;
-
-    // TODO: This function is very heavy. We should try to do some caching here, maybe in the
```
Comment (Jongy) [inline examples/cpp/pyperf/PyPerfNativeStackTrace.cc:74]: Let's keep the TODO next to `unw_get_proc_name`?

File: examples/cpp/pyperf/PyPerfNativeStackTrace.cc [hunk 16]
```diff
@@ -65,26 +84,29 @@ NativeStackTrace::NativeStackTrace(uint32_t pid, const unsigned char *raw_stack,
...
-      } else {
-        this->symbols.push_back(std::string(sym));
-      }
-      free(realname);
+    unw_word_t ip;
+    unw_get_reg(&cursor, UNW_REG_IP, &ip);
+    if (procSymbols && procSymbols->resolve_addr(ip, &resolved_symbol, true)) {
+        snprintf(buf, buf_size, "%s (%s)", resolved_symbol.demangle_name, resolved_symbol.module);
+        this->symbols.push_back(std::string(buf));
+    } else if (!unw_get_proc_name(&cursor, sym, sizeof(sym), &offset)) {
```
Comment (Jongy) [inline examples/cpp/pyperf/PyPerfNativeStackTrace.cc:99]:
```suggestion
    } else if (!unw_get_proc_name(&cursor, buf, sizeof(buf), &offset)) {
```

for what it's worth, we already have a larger buffer, let's use it and avoid getting truncated names :)

Comment (Jongy) [reply examples/cpp/pyperf/PyPerfNativeStackTrace.cc:99]: You can remove `sym` altogether I guess

Comment (Jongy) [inline examples/cpp/pyperf/PyPerfNativeStackTrace.cc:99]: Code used to store the result of `unw_get_proc_name` in `res` and use it in the following `unw_sterror`, and you no longer store `res` here. Let's keep it (you can use `if (!(res = unw_get_proc_name(...)))`)

File: examples/cpp/pyperf/PyPerfNativeStackTrace.cc [hunk 17]
```diff
@@ -200,5 +222,35 @@ bool NativeStackTrace::error_occured() const {
...
+    procSymbolsCache.erase(it);
+  }
+}
+
+void NativeStackTrace::enable_dso_reporting() {
+  NativeStackTrace::insert_dso_name = true;
+}
+
+ProcSyms* NativeStackTrace::get_proc_symbols(uint32_t pid) {
+  static struct bcc_symbol_option symbol_options = {
```
Comment (Jongy) [inline examples/cpp/pyperf/PyPerfNativeStackTrace.cc:237]: Since it's not `const` - I think it's best to define it without `static`, so in case the function does modify it in any way, it will not affect subsequent calls.

**Review by Jongy — CHANGES_REQUESTED** [review]
Done reviewing. Logic appears sound to me. All that's left is small comments

File: examples/cpp/pyperf/PyPerfNativeStackTrace.cc (hunk 13 above)
Comment (marcin-ol) [reply examples/cpp/pyperf/PyPerfNativeStackTrace.cc]: 👍

File: examples/cpp/pyperf/PyPerfNativeStackTrace.cc (hunk 14 above)
Comment (marcin-ol) [reply examples/cpp/pyperf/PyPerfNativeStackTrace.cc]: Agreed.

File: examples/cpp/pyperf/PyPerfNativeStackTrace.cc (hunk 15 above)
Comment (marcin-ol) [reply examples/cpp/pyperf/PyPerfNativeStackTrace.cc:74]: Sure.

File: examples/cpp/pyperf/PyPerfNativeStackTrace.cc (hunk 16 above)
Comment (marcin-ol) [reply examples/cpp/pyperf/PyPerfNativeStackTrace.cc]: 👍

Comment (marcin-ol) [reply examples/cpp/pyperf/PyPerfNativeStackTrace.cc]: Thanks for spotting it.

File: examples/cpp/pyperf/PyPerfNativeStackTrace.cc (hunk 17 above)
Comment (marcin-ol) [reply examples/cpp/pyperf/PyPerfNativeStackTrace.cc]: Sure.

**Review by Jongy — APPROVED** [review]
Tests pass in gProfiler PR which uses this commit.

**Comment by Jongy** [comment]
I pushed tag: v1.2.4
